File: 863-all-nodes-distance-k-in-binary-tree/863. All Nodes Distance K in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        adj = defaultdict(list)

        def dfs(node, prev):
            print(node.val, prev)
            if prev is not None:
                adj[node.val].append(prev)
            if node.left:
                adj[node.val].append(node.left.val)
                dfs(node.left, node.val)
            if node.right:
                adj[node.val].append(node.right.val)
                dfs(node.right, node.val)
        dfs(root, None)
        print(adj)
        queue = deque([target.val])

        visited = set()
        visited.add(target.val)
        for j in range(k):
            for i in range(len(queue)):
                node = queue.popleft()
                for nei in adj[node]:
                    if nei not in visited:
                        visited.add(nei)
                        queue.append(nei)
            print(queue)

        return list(queue)

File: 863-all-nodes-distance-k-in-binary-tree/863. All Nodes Distance K in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    struct KeyHash {
        size_t operator()(const TreeNode* t) const {
            return hash<int>()(t->val);
        }
    };
    void dfs(TreeNode* node, unordered_map<TreeNode*, vector<TreeNode*>, KeyHash>& graph) {
        if (node == nullptr) return;
        if (node->left) {
            graph[node].push_back(node->left);
            graph[node->left].push_back(node);
            dfs(node->left, graph);
        }
        if (node->right) {
            graph[node].push_back(node->right);
            graph[node->right].push_back(node);
            dfs(node->right, graph);
        }
    }
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_map<TreeNode*, vector<TreeNode*>, KeyHash> graph;
        dfs(root, graph);
        queue<TreeNode*> q;
        q.push(target);
        unordered_set<TreeNode*> seen;
        int steps = 0;
        vector<int> res;
        seen.insert(target);
        while (!q.empty()) {
            int q_size = q.size();
            for (int i = 0; i < q_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                if (steps == k) {
                    res.push_back(node->val);
                    continue;
                }
                for (auto nei : graph[node]) {
                    if (!seen.count(nei)) {
                        seen.insert(nei);
                        q.push(nei);
                    }
                }
            }
            steps++;
        }
        return res;
    }
};